internal/plog/level_test.go

```diff
@@ -4,14 +4,169 @@
+  "caller": "%s/level_test.go:%d$plog.TestFormat",
+  "msg": "wee",
+  "a": "b",
+  "error": "invalid log level, valid choices are the empty string, info, debug, trace and all",
+  "logger": "burrito"
+}`, wd, startLogLine+2+13), scanner.Text())
+
+	Logr().V(klogLevelWarning).Info("hey") // note that this fails to set the custom warning key because it is not via plog
+	require.True(t, scanner.Scan())
+	require.NoError(t, scanner.Err())
+	require.JSONEq(t, fmt.Sprintf(`
+{
+  "level": "info",
+  "ts": "2022-11-21T23:37:26.953313Z",
+  "caller": "%s/level_test.go:%d$plog.TestFormat",
+  "msg": "hey"
+}`, wd, startLogLine+2+13+14), scanner.Text())
+
+	Warning("bad stuff") // note that this sets the custom warning key because it is via plog
+	require.True(t, scanner.Scan())
+	require.NoError(t, scanner.Err())
+	require.JSONEq(t, fmt.Sprintf(`
+{
+  "level": "info",
+  "ts": "2022-11-21T23:37:26.953313Z",
+  "caller": "%s/level_test.go:%d$plog.TestFormat",
+  "msg": "bad stuff",
+  "warning": "true"
+}`, wd, startLogLine+2+13+14+11), scanner.Text())
+
+	DebugErr("something happened", errInvalidLogFormat, "an", "item")
+	require.True(t, scanner.Scan())
+	require.NoError(t, scanner.Err())
+	require.JSONEq(t, fmt.Sprintf(`
+{
+  "level": "debug",
+  "ts": "2022-11-21T23:37:26.953313Z",
+  "caller": "%s/level_test.go:%d$plog.TestFormat",
+  "msg": "something happened",
+  "error": "invalid log format, valid choices are the empty string, json and text",
+  "an": "item"
+}`, wd, startLogLine+2+13+14+11+12), scanner.Text())
+
+	Trace("should not be logged", "for", "sure")
+	require.Empty(t, buf.String())
+
+	Logr().V(klogLevelAll).Info("also should not be logged", "open", "close")
+	require.Empty(t, buf.String())
+
+	err = ValidateAndSetLogLevelAndFormatGlobally(ctx, LogSpec{Level: LevelDebug, Format: FormatText},
+		zap.WithClock(&clockAdapter{clock: fakeClock}),
+		zap.WrapCore(func(core zapcore.Core) zapcore.Core {
+			// this is some dark magic to set a private field
+			outField := reflect.ValueOf(core).Elem().FieldByName("out")
+			outPointer := (*zapcore.WriteSyncer)(unsafe.Pointer(outField.UnsafeAddr()))
+			*outPointer = ws
+			return core
+		}),
+	)
+	require.NoError(t, err)
+
+	DebugErr("something happened", errInvalidLogFormat, "an", "item")
+	require.True(t, scanner.Scan())
+	require.NoError(t, scanner.Err())
+	require.Equal(t, fmt.Sprintf(`2022-11-21T23:37:26.953313Z	debug	%s/level_test.go:%d$plog.TestFormat	something happened	{"error": "invalid log format, valid choices are the empty string, json and text", "an": "item"}`,
+		wd, startLogLine+2+13+14+11+12+31), scanner.Text())
+
+	Logr().WithName("burrito").Error(errInvalidLogLevel, "wee", "a", "b")
+	require.True(t, scanner.Scan())
+	require.NoError(t, scanner.Err())
+	require.Equal(t, fmt.Sprintf(`2022-11-21T23:37:26.953313Z	error	burrito	%s/level_test.go:%d$plog.TestFormat	wee	{"a": "b", "error": "invalid log level, valid choices are the empty string, info, debug, trace and all"}`,
```

Comment:
This shows the `text` format of the exact same log statement I commented on earlier for the JSON format.  Honestly I question offering this as a choice because I cannot imagine anyone would find this better than the JSON output.  Note that while the key/vals are encoded as JSON here, we can override that use a different format if we so choose (we just have to write the code for said custom format).  We cannot use the old `klog` format while keeping the really robust zap config (it is all hidden inside a global/private logger).  In particular, with the zap config it would be easy in the future for us to redact logs from particular packages, meaning we could prevent secret disclosure (at high log level) caused by the Kube library code we use.

My proposal here would be to not offer any log format config, and simply always use JSON.  Then in the future we can easily rollout enhancements to the logging such as redaction, control over where the logs are sent, when stack traces are emitted, etc.